paimon-core/src/main/java/org/apache/paimon/manifest/IndexManifestFileHandler.java

```diff
@@ -0,0 +1,150 @@
+        List<IndexManifestEntry> entries =
+                previousIndexManifest == null
+                        ? new ArrayList<>()
+                        : indexManifestFile.read(previousIndexManifest);
+        Pair<List<IndexManifestEntry>, List<IndexManifestEntry>> previous =
+                separateIndexEntries(entries);
+        Pair<List<IndexManifestEntry>, List<IndexManifestEntry>> current =
+                separateIndexEntries(newIndexFiles);
+
+        // Step1: get the hash index files;
+        List<IndexManifestEntry> indexEntries =
+                getIndexManifestFileCombine(HASH_INDEX)
+                        .combine(previous.getLeft(), current.getLeft());
+
+        // Step2: get the dv index files;
+        indexEntries.addAll(
+                getIndexManifestFileCombine(DELETION_VECTORS_INDEX)
+                        .combine(previous.getRight(), current.getRight()));
+
+        return indexManifestFile.writeWithoutRolling(indexEntries);
+    }
+
+    private Pair<List<IndexManifestEntry>, List<IndexManifestEntry>> separateIndexEntries(
+            List<IndexManifestEntry> indexFiles) {
+        List<IndexManifestEntry> hashEntries = new ArrayList<>();
+        List<IndexManifestEntry> dvEntries = new ArrayList<>();
+        for (IndexManifestEntry entry : indexFiles) {
+            String indexType = entry.indexFile().indexType();
+            if (indexType.equals(DELETION_VECTORS_INDEX)) {
+                dvEntries.add(entry);
+            } else if (indexType.equals(HASH_INDEX)) {
+                hashEntries.add(entry);
+            } else {
+                throw new IllegalArgumentException("Can't recognize this index type: " + indexType);
+            }
+        }
+        return Pair.of(hashEntries, dvEntries);
+    }
+
+    private IndexManifestFileCombiner getIndexManifestFileCombine(String indexType) {
+        if (DELETION_VECTORS_INDEX.equals(indexType) && BucketMode.BUCKET_UNAWARE == bucketMode) {
+            return new UnawareBucketCombiner();
+        } else {
+            return new CommonBucketCombiner();
+        }
+    }
+
+    interface IndexManifestFileCombiner {
+        List<IndexManifestEntry> combine(
+                List<IndexManifestEntry> prevIndexFiles, List<IndexManifestEntry> newIndexFiles);
+    }
+
+    /**
+     * We combine the previous and new index files by the file name. This is only used for tables
+     * with UnawareBucket.
+     */
+    static class UnawareBucketCombiner implements IndexManifestFileCombiner {
+
+        @Override
+        public List<IndexManifestEntry> combine(
+                List<IndexManifestEntry> prevIndexFiles, List<IndexManifestEntry> newIndexFiles) {
+            Map<String, IndexManifestEntry> indexEntries = new LinkedHashMap<>();
+            for (IndexManifestEntry entry : prevIndexFiles) {
+                if (entry.kind() == FileKind.ADD) {
```

Comment:
prevIndexFiles should has not delete files. here just checkArguement?
(maybe you can assert outside)

Reply:
yes.  all the `prevIndexFiles` should be FileKind.ADD.